mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala

```diff
@@ -276,44 +277,53 @@ object MatrixFactorizationModel extends Loader[MatrixFactorizationModel] {
       num: Int): RDD[(Int, Array[(Int, Double)])] = {
     val srcBlocks = blockify(rank, srcFeatures)
     val dstBlocks = blockify(rank, dstFeatures)
-    val ratings = srcBlocks.cartesian(dstBlocks).flatMap {
-      case ((srcIds, srcFactors), (dstIds, dstFactors)) =>
-        val m = srcIds.length
-        val n = dstIds.length
-        val ratings = srcFactors.transpose.multiply(dstFactors)
-        val output = new Array[(Int, (Int, Double))](m * n)
-        var k = 0
-        ratings.foreachActive { (i, j, r) =>
-          output(k) = (srcIds(i), (dstIds(j), r))
-          k += 1
+    /**
+     * Use dot to replace blas 3 gemm is the key approach to improve efficiency.
+     * By this change, we can get the topK elements of each block to reduce the GC time.
+     * Comparing with BLAS.dot, hand-written dot is high efficiency.
+     */
+    val ratings = srcBlocks.cartesian(dstBlocks).flatMap { case (srcIter, dstIter) =>
+      val m = srcIter.size
+      val n = math.min(dstIter.size, num)
+      val output = new Array[(Int, (Int, Double))](m * n)
+      var j = 0
+      srcIter.foreach { case (srcId, srcFactor) =>
+        val pq = new BoundedPriorityQueue[(Int, Double)](n)(Ordering.by(_._2))
```

Comment:
Here we are creating O(blocksize) PQ objects. But we could have only one, and reuse it. We could create it outside the `foreach` calls and call `clear` after using it. Not really sure if it will make a large difference, but could further reduce GC